<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>SPCompute</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">SPCompute</h1>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SPCompute)</span></code></pre></div>
<div id="section-1-compute-powersample-size-for-binary-traits" class="section level2">
<h2>Section 1: Compute power/sample size for binary traits</h2>
<p>When the trait of interest is binary, we assume the following
logistic regression model: <span class="math display">\[\log\bigg(\frac{\text{P}(Y_i=1|X)}{1-\text{P}(Y_i=1|X)}\bigg)
= X\beta,\]</span> where the design matrix contains a column of <span class="math inline">\(1&#39;s\)</span> for the intercept, a column of
genotypes <span class="math inline">\(G\)</span> and a column for
non-genetic covariate <span class="math inline">\(E\)</span>
(optional).</p>
<p>The regression parameter vector <span class="math inline">\(\beta\)</span> contains <span class="math inline">\(\beta_0\)</span>, <span class="math inline">\(\beta_G\)</span> and <span class="math inline">\(\beta_E\)</span>, respectively represent intercept
parameter, genetic effect and covariate effect.</p>
<p>To compute power or sample size, the user will need to specify the
following information:</p>
<ul>
<li><code>preva</code>, the prevalence rate of the disease in the
population, defined as <span class="math inline">\(\text{P}(Y=1)\)</span>.</li>
<li><code>betaG</code>, the true effect size of genetic effect.</li>
<li><code>pG</code>, the minor allele frequency of the SNP.</li>
</ul>
<p>If there exists non-genetic covariate <span class="math inline">\(E\)</span> in the model, the user will also need
to specify the following parameters:</p>
<ul>
<li><code>betaE</code>, the true effect size of non-genetic covariate
effect.</li>
<li><code>gammaG</code>, the parameter that specifies the dependency
between <span class="math inline">\(E\)</span> and <span class="math inline">\(G\)</span>.</li>
</ul>
<p>If the non-genetic covariate <span class="math inline">\(E\)</span>
is binary, the following marginal information on <span class="math inline">\(E\)</span> should be specified:</p>
<ul>
<li><p><code>pE</code>, the population prevalence rate of <span class="math inline">\(E\)</span>, defined as <span class="math inline">\(\text{P}(E=1)\)</span>. Otherwise if it is
continuous, the required marginal information should be:</p></li>
<li><p><code>muE</code>, the population mean of <span class="math inline">\(E\)</span>.</p></li>
<li><p><code>sigmaE</code>, the population SD of <span class="math inline">\(E\)</span>.</p></li>
</ul>
<p>These parameters should be summarized into a list, with appropriate
names, such as the following when covariate is binary:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>para <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">preva =</span> <span class="fl">0.2</span>, <span class="at">pG =</span> <span class="fl">0.1</span>, <span class="at">betaG =</span> <span class="fl">0.1</span>, <span class="at">betaE =</span> <span class="fl">0.3</span>, <span class="at">pE =</span> <span class="fl">0.3</span>, <span class="at">gammaG =</span> <span class="dv">0</span>)</span></code></pre></div>
<p>To compute power given a sample size <code>n</code>, the user can use
the function <code>Compute_Size</code>, after specifying the argument
for:</p>
<ul>
<li><code>parameters</code>, a list of true parameter values defined as
above.</li>
<li><code>n</code>, the given sample size</li>
<li><code>covariate</code>, the type of covariate, should be “binary”,
“continuous” or “none”.</li>
<li><code>mode</code>, the genetic mode, should be “additive”,
“dominant” or “recessive”.</li>
<li><code>alpha</code>, the significance level.</li>
<li><code>method</code>, the method used to do the computation. Should
be “semi-sim” (faster for large sample) or “expand” (better for smaller
sample).</li>
</ul>
<p>For example:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Compute_Power</span>(<span class="at">parameters =</span> para, <span class="at">n =</span> <span class="fl">2e4</span>, <span class="at">covariate =</span> <span class="st">&quot;binary&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;additive&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>, <span class="at">method =</span> <span class="st">&quot;semi-sim&quot;</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.6908458</span></span></code></pre></div>
<p>or:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Compute_Power</span>(<span class="at">parameters =</span> para, <span class="at">n =</span> <span class="fl">2e4</span>, <span class="at">covariate =</span> <span class="st">&quot;binary&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;additive&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>, <span class="at">method =</span> <span class="st">&quot;expand&quot;</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.6884145</span></span></code></pre></div>
<p>Similarly, to compute the required sample size to achieve a certain
power, one just needs to change the argument <code>n</code> to
<code>PowerAim</code>, which defines the target power, and uses the
function <code>Compute_Size</code>:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">Compute_Size</span>(<span class="at">parameters =</span> para, <span class="at">PowerAim =</span> <span class="fl">0.8</span>, <span class="at">covariate =</span> <span class="st">&quot;binary&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;additive&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>, <span class="at">method =</span> <span class="st">&quot;semi-sim&quot;</span>))</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 25978</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">Compute_Size</span>(<span class="at">parameters =</span> para, <span class="at">PowerAim =</span> <span class="fl">0.8</span>, <span class="at">covariate =</span> <span class="st">&quot;binary&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;additive&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>, <span class="at">method =</span> <span class="st">&quot;expand&quot;</span>))</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 26128</span></span></code></pre></div>
<p>or if the genetic mode is dominant:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">Compute_Size</span>(<span class="at">parameters =</span> para, <span class="at">PowerAim =</span> <span class="fl">0.8</span>, <span class="at">covariate =</span> <span class="st">&quot;binary&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;dominant&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>, <span class="at">method =</span> <span class="st">&quot;semi-sim&quot;</span>))</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 30872</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">Compute_Size</span>(<span class="at">parameters =</span> para, <span class="at">PowerAim =</span> <span class="fl">0.8</span>, <span class="at">covariate =</span> <span class="st">&quot;binary&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;dominant&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>, <span class="at">method =</span> <span class="st">&quot;expand&quot;</span>))</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 30863</span></span></code></pre></div>
<p>If one wants to have more accurate estimate of sample size or power
when using the method <code>semi-sim</code>, the parameter
<code>B</code> can be set to larger value, which will take longer
run-time:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">Compute_Size</span>(<span class="at">parameters =</span> para, <span class="at">PowerAim =</span> <span class="fl">0.8</span>, <span class="at">covariate =</span> <span class="st">&quot;binary&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;dominant&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>, <span class="at">method =</span> <span class="st">&quot;semi-sim&quot;</span>, <span class="at">B =</span> <span class="fl">5e5</span>))</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 30817</span></span></code></pre></div>
<p>Unlike the case to compute power, when computing sample size, it is
always recommended to use the method <code>semi-sim</code>, since the
method <code>expand</code> will not work when the sample size is
extremely small, and will work at a much slower speed when the sample
size is extremely large.</p>
</div>
<div id="section-2-compute-powersample-size-for-continuous-traits" class="section level2">
<h2>Section 2: Compute power/sample size for continuous traits</h2>
<p>When the trait of interest is continuous, we assume the following
linear regression model: <span class="math display">\[Y = X\beta +
\epsilon,\]</span> where the noise <span class="math inline">\(\epsilon\sim N(0,\sigma_\epsilon^2)\)</span>.</p>
<p>To compute power or sample size, the procedures will be always the
same as in the binary case, except that the parameter <code>preva</code>
will be replaced by the set of parameters:</p>
<ul>
<li><code>TraitMean</code>, specifying the population mean of the
continuous trait, i.e. <span class="math inline">\(\mathbb{E}(Y)\)</span>.</li>
<li><code>TraitSD</code>, specifying the population standard deviation
of the continuous trait.</li>
</ul>
<p>The R function <code>Compute_Size</code> or
<code>Compute_Power</code> will then compute quantities such as <span class="math inline">\(\sigma_\epsilon\)</span> using these marginal
information automatically.</p>
<p>Alternatively, the user may also inputs the values of <span class="math inline">\(\sigma_\epsilon\)</span> directly instead of
inputting the value of <code>TraitSD</code>, by replacing
<code>TraitSD</code> with</p>
<ul>
<li><code>ResidualSD</code>, the value of <span class="math inline">\(\sigma_\epsilon\)</span> in the model.</li>
</ul>
<p>Now the computation can be proceeded by specifying
<code>response = &quot;continuous&quot;</code>:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>para <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">TraitMean =</span> <span class="dv">3</span>, <span class="at">TraitSD =</span> <span class="dv">1</span>, <span class="at">pG =</span> <span class="fl">0.1</span>, <span class="at">betaG =</span> <span class="fl">0.1</span>, <span class="at">betaE =</span> <span class="fl">0.3</span>, <span class="at">pE =</span> <span class="fl">0.3</span>, <span class="at">gammaG =</span> <span class="dv">0</span>)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Compute_Power</span>(<span class="at">parameters =</span> para, <span class="at">n =</span> <span class="fl">5e3</span>, <span class="at">covariate =</span> <span class="st">&quot;binary&quot;</span>, <span class="at">response =</span> <span class="st">&quot;continuous&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;additive&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.8580452</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">Compute_Size</span>(<span class="at">parameters =</span> para, <span class="at">PowerAim =</span> <span class="fl">0.8</span>, <span class="at">response =</span> <span class="st">&quot;continuous&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;additive&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>))</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 4270</span></span></code></pre></div>
<p>Or:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>para <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">TraitMean =</span> <span class="dv">3</span>, <span class="at">ResidualSD =</span> <span class="dv">1</span>, <span class="at">pG =</span> <span class="fl">0.1</span>, <span class="at">betaG =</span> <span class="fl">0.1</span>, <span class="at">betaE =</span> <span class="fl">0.3</span>, <span class="at">pE =</span> <span class="fl">0.3</span>, <span class="at">gammaG =</span> <span class="dv">0</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Compute_Power</span>(<span class="at">parameters =</span> para, <span class="at">n =</span> <span class="fl">5e3</span>, <span class="at">covariate =</span> <span class="st">&quot;binary&quot;</span>, <span class="at">response =</span> <span class="st">&quot;continuous&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;additive&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.8508388</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">Compute_Size</span>(<span class="at">parameters =</span> para, <span class="at">PowerAim =</span> <span class="fl">0.8</span>, <span class="at">response =</span> <span class="st">&quot;continuous&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;additive&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>))</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 4360</span></span></code></pre></div>
<p>Note that for continuous trait, the value of <code>TraitMean</code>
is only used to do parameter conversion, which will not affect the
result of power or sample size computation. So if the purpose is not to
compute the converted parameter such as <span class="math inline">\(\beta_0\)</span>, the value of
<code>TraitMean</code> can be set to arbitrary numeric value, as shown
in the following example:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>para <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">TraitMean =</span> <span class="dv">3</span>, <span class="at">ResidualSD =</span> <span class="dv">1</span>, <span class="at">pG =</span> <span class="fl">0.1</span>, <span class="at">betaG =</span> <span class="fl">0.1</span>, <span class="at">betaE =</span> <span class="fl">0.3</span>, <span class="at">pE =</span> <span class="fl">0.3</span>, <span class="at">gammaG =</span> <span class="dv">0</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Compute_Power</span>(<span class="at">parameters =</span> para, <span class="at">n =</span> <span class="fl">5e3</span>, <span class="at">covariate =</span> <span class="st">&quot;binary&quot;</span>, <span class="at">response =</span> <span class="st">&quot;continuous&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;additive&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.8508388</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">Compute_Size</span>(<span class="at">parameters =</span> para, <span class="at">PowerAim =</span> <span class="fl">0.8</span>, <span class="at">response =</span> <span class="st">&quot;continuous&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;additive&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>))</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 4360</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>para <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">TraitMean =</span> <span class="dv">30000</span>, <span class="at">ResidualSD =</span> <span class="dv">1</span>, <span class="at">pG =</span> <span class="fl">0.1</span>, <span class="at">betaG =</span> <span class="fl">0.1</span>, <span class="at">betaE =</span> <span class="fl">0.3</span>, <span class="at">pE =</span> <span class="fl">0.3</span>, <span class="at">gammaG =</span> <span class="dv">0</span>)</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="fu">Compute_Power</span>(<span class="at">parameters =</span> para, <span class="at">n =</span> <span class="fl">5e3</span>, <span class="at">covariate =</span> <span class="st">&quot;binary&quot;</span>, <span class="at">response =</span> <span class="st">&quot;continuous&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;additive&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>)</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.8508388</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">Compute_Size</span>(<span class="at">parameters =</span> para, <span class="at">PowerAim =</span> <span class="fl">0.8</span>, <span class="at">response =</span> <span class="st">&quot;continuous&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;additive&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>))</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 4360</span></span></code></pre></div>
<p>Similarly, when the covariate <span class="math inline">\(E\)</span>
and the SNP <span class="math inline">\(G\)</span> are independent
(i.e. <code>gammaG = 0</code>), power or sample size computation will
not depend on the covariate information at all:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>para <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">TraitMean =</span> <span class="dv">0</span>, <span class="at">ResidualSD =</span> <span class="dv">1</span>, <span class="at">pG =</span> <span class="fl">0.1</span>, <span class="at">betaG =</span> <span class="fl">0.1</span>, <span class="at">betaE =</span> <span class="dv">1000</span>, <span class="at">pE =</span> <span class="fl">0.5</span>, <span class="at">gammaG =</span> <span class="dv">0</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Compute_Power</span>(<span class="at">parameters =</span> para, <span class="at">n =</span> <span class="fl">5e3</span>, <span class="at">covariate =</span> <span class="st">&quot;binary&quot;</span>, <span class="at">response =</span> <span class="st">&quot;continuous&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;additive&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.8508388</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">Compute_Size</span>(<span class="at">parameters =</span> para, <span class="at">PowerAim =</span> <span class="fl">0.8</span>, <span class="at">response =</span> <span class="st">&quot;continuous&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;additive&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>))</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 4360</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>para <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">TraitMean =</span> <span class="dv">0</span>, <span class="at">ResidualSD =</span> <span class="dv">1</span>, <span class="at">pG =</span> <span class="fl">0.1</span>, <span class="at">betaG =</span> <span class="fl">0.1</span>, <span class="at">betaE =</span> <span class="fl">0.001</span>, <span class="at">pE =</span> <span class="fl">0.1</span>, <span class="at">gammaG =</span> <span class="dv">0</span>)</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="fu">Compute_Power</span>(<span class="at">parameters =</span> para, <span class="at">n =</span> <span class="fl">5e3</span>, <span class="at">covariate =</span> <span class="st">&quot;binary&quot;</span>, <span class="at">response =</span> <span class="st">&quot;continuous&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;additive&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>)</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.8508388</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">Compute_Size</span>(<span class="at">parameters =</span> para, <span class="at">PowerAim =</span> <span class="fl">0.8</span>, <span class="at">response =</span> <span class="st">&quot;continuous&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;additive&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>))</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 4360</span></span></code></pre></div>
</div>
<div id="section-3-powersample-size-computation-with-multiple-covariates" class="section level2">
<h2>Section 3: Power/Sample Size Computation with Multiple
Covariates</h2>
<p>When there are multiple covariates (<span class="math inline">\(E_1\)</span> and <span class="math inline">\(E_2\)</span>) simultaneously affecting the trait
<span class="math inline">\(Y\)</span>, we assume the following (glm)
model: <span class="math display">\[g_1[\mathbb{E}(Y|E_1,E_2,G)] =
\beta_0 + \beta_GG + \beta_{E_1}E_1 + \beta_{E_2}E_2.\]</span></p>
<p>The dependency between <span class="math inline">\(E_1, E_2\)</span>
and <span class="math inline">\(G\)</span> are specified through the
following (nested) second stage regressions: <span class="math display">\[\begin{equation}
\begin{aligned}
g_2[\mathbb{E}(E_1|G)] &amp;= \gamma_{01} + \gamma_{G_1}G\\
g_3[\mathbb{E}(E_2|G,E_1)] &amp;= \gamma_{02} + \gamma_{G_2}G + \gamma_E
E_1.\\
\end{aligned}
\end{equation}\]</span></p>
<p>The covariates and the trait could be either continuous or binary,
where the link functions <span class="math inline">\(g_1,g_2\)</span>
and <span class="math inline">\(g_3\)</span> correspond to linear and
logistic regression respectively as described in previous sections.</p>
<p>For example when <span class="math inline">\(Y\)</span> is binary,
<span class="math inline">\(E_1\)</span> is binary (with <span class="math inline">\(\beta_{E_1} = 0.3,\  \text{P}(E_1=1)=0.3\)</span>)
and <span class="math inline">\(E_2\)</span> is continuous (with <span class="math inline">\(\beta_{E_2} = 0.2,\ \mathbb{E}(E_2) = 0, \
\text{Var}(E_2) = 0.4\)</span>), the power can be computed as:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>para <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">preva =</span> <span class="fl">0.2</span>, <span class="at">pG =</span> <span class="fl">0.1</span>, <span class="at">betaG =</span> <span class="fl">0.1</span>, <span class="at">betaE =</span> <span class="fu">c</span>(<span class="fl">0.3</span>, <span class="fl">0.2</span>), </span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">pE =</span> <span class="fl">0.3</span>, <span class="at">gammaG =</span> <span class="fu">c</span>(<span class="fl">0.15</span>,<span class="fl">0.25</span>), <span class="at">gammaE =</span> <span class="fl">0.1</span>,</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">muE =</span> <span class="dv">0</span>, <span class="at">sigmaE =</span> <span class="fu">sqrt</span>(<span class="fl">0.4</span>))</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="fu">Compute_Power_multi</span>(<span class="at">parameters =</span> para, <span class="at">n =</span> <span class="dv">3000</span>, <span class="at">mode =</span> <span class="st">&quot;additive&quot;</span>,</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">covariate =</span> <span class="fu">c</span>(<span class="st">&quot;binary&quot;</span>, <span class="st">&quot;continuous&quot;</span>), <span class="at">response =</span> <span class="st">&quot;binary&quot;</span>)</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.159002</span></span></code></pre></div>
<p>In other words, this correspond to the models: <span class="math display">\[\begin{equation}
\begin{aligned}
\log\bigg(\frac{\text{P}(Y=1|E_1,E_2,G)}{1-\text{P}(Y=1|E_1,E_2,G)}\bigg)
&amp;= \beta_0 + 0.1 G + 0.3 E_1 + 0.2 E_2 \\
\log\bigg(\frac{\text{P}(E_1=1|G)}{1-\text{P}(E_1=1|G)}\bigg) &amp;=
\gamma_{01} + 0.15 G\\
\mathbb{E}(E_2|G,E_1) &amp;= \gamma_{02} + 0.25 G + 0.1 E_1,\\
\end{aligned}
\end{equation}\]</span> where all the remaining parameters <span class="math inline">\(\bigg(\beta_0, \gamma_{01}, \gamma_{02},
\text{Var}[E_2|G,E_1] \bigg)\)</span> are automatically solved using the
marginal information on <span class="math inline">\(Y\)</span>, <span class="math inline">\(E_1\)</span>, <span class="math inline">\(E_2\)</span> and <span class="math inline">\(G\)</span>.</p>
<p>Note that when <code>gammaE</code> is unspecified, by the default it
is assumed the two covariates are conditionally independent given <span class="math inline">\(G\)</span>.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
